File: lib/src/domain/entities/svg_motion_path.dart
import 'package:meta/meta.dart';

/// CSS Motion Path attached to a node via inline `style` (CSS Motion Path
/// Module Level 1 — `offset-path`, `offset-rotate`).
///
/// The associated animation of [offset-distance] drives the node along
/// [pathData]; [rotate] controls whether the node also rotates to match the
/// path tangent at each sampled point.
///
/// Concrete translate/rotate keyframes are produced later by
/// `MotionPathResolver`, which samples the path once the `SvgDocument` is
/// available (the CSS parser cannot see geometry on the node).
@immutable
class SvgMotionPath {
  const SvgMotionPath({
    required this.pathData,
    this.rotate = const SvgMotionRotate.auto(),
  });

  /// Raw SVG path `d` string as declared inside `offset-path: path('...')`.
  /// Quotes stripped; whitespace preserved.
  final String pathData;

  final SvgMotionRotate rotate;
}

/// `offset-rotate` value. `auto` → follow path tangent; `reverse` → tangent
/// flipped 180°; `fixed(deg)` → constant angle (tangent ignored, e.g.
/// `offset-rotate: 0deg` keeps the sprite upright).
@immutable
class SvgMotionRotate {
  const SvgMotionRotate.auto()
      : kind = SvgMotionRotateKind.auto,
        angleDeg = 0;
  const SvgMotionRotate.reverse()
      : kind = SvgMotionRotateKind.reverse,
        angleDeg = 0;
  const SvgMotionRotate.fixed(this.angleDeg)
      : kind = SvgMotionRotateKind.fixed;

  final SvgMotionRotateKind kind;
  final double angleDeg;
}

enum SvgMotionRotateKind { auto, reverse, fixed }

File: native/anim_svg_core/src/domain/svg_motion_path.rs
//! Port of `lib/src/domain/entities/svg_motion_path.dart`.

use serde::Serialize;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum SvgMotionRotateKind {
    Auto,
    Reverse,
    Fixed,
}

/// CSS `offset-rotate`. `Fixed(deg)` keeps the node upright at a constant
/// angle; `Auto` / `Reverse` follow the path tangent.
#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SvgMotionRotate {
    pub kind: SvgMotionRotateKind,
    #[serde(default)]
    pub angle_deg: f64,
}

impl SvgMotionRotate {
    pub const AUTO: Self = Self {
        kind: SvgMotionRotateKind::Auto,
        angle_deg: 0.0,
    };
    pub const REVERSE: Self = Self {
        kind: SvgMotionRotateKind::Reverse,
        angle_deg: 0.0,
    };
    pub const fn fixed(angle_deg: f64) -> Self {
        Self {
            kind: SvgMotionRotateKind::Fixed,
            angle_deg,
        }
    }
}

impl Default for SvgMotionRotate {
    fn default() -> Self {
        Self::AUTO
    }
}

/// CSS Motion Path (`offset-path: path('...')` + `offset-rotate: ...`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SvgMotionPath {
    pub path_data: String,
    #[serde(default)]
    pub rotate: SvgMotionRotate,
}
